mongo/options/insertoptions.go

```diff
@@ -19,20 +19,36 @@ type InsertOneOptions struct {
 	Comment interface{}
 }
 
+// InsertOneOptions represents functional options that configure an InsertOneArgs.
+type InsertOneOptions struct {
+	Opts []func(*InsertOneArgs) error
+}
```

Comment:
@matthewdale @qingyang-hu I just realized that one breaking change for the current design is that a user will no longer be able to decode bytes into an options type:

```go
package main

import (
	"encoding/json"

	"go.mongodb.org/mongo-driver/mongo/options"
)

func main() {
	bytes := []byte(`{"comment": "hi"}`)

	ioo := options.InsertOneOptions{}
	if err := json.Unmarshal(bytes, &ioo); err != nil { // Results in nothing
		panic(err)
	}

	// ...
}
```

If we want to keep the current suggested pattern, we could create a convenience API to construct options from arguments:

```go
func InsertOneFromArgs(args *InsertOneArgs) *InsertOneOptions {
	opts := &InsertOneOptions{}
	if args == nil {
		return opts
	}

	if args.BypassDocumentValidation != nil {
		opts.SetBypassDocumentValidation(*args.BypassDocumentValidation)
	}

	if args.Comment != nil {
		opts.SetComment(args.Comment)
	}

	return opts
}
```

There is at least one legitimate use case for this in the Go Driver, specifically in the [`executeListSearchIndexes` ](https://github.com/mongodb/mongo-go-driver/blob/master/internal/integration/unified/collection_operation_execution.go#L1138) in the unified spec runner . Though this pattern might actually be an antipattern when compared with what is typically done with options, such as [`executeAggregate`](https://github.com/mongodb/mongo-go-driver/blob/master/internal/integration/unified/collection_operation_execution.go#L52).

Should we continue moving forward with the existing pattern?

Reply:
I think we can create types in the unified test runner package(s) that implement `mongo.Options` for the specific operation type, but can decode the args from JSON.

E.g.
```go
type testInsertOneOpts struct {
	data []byte
}

func (opts *testInsertOneOpts) ArgsSetters() []func(*options.InsertOneArgs) error {
	return []func(*options.InsertOneArgs) error{
		func(args *options.InsertOneArgs) error {
			return json.Unmarshal(opts.data, args)
		},
	}
}
```
Then we can pass that to the operation function:
```go
coll.InsertOne(
	context.Background(),
	testDoc,
	&testInsertOneOpts{data: testArgsBytes})
```